privacy: Fix more (potential) issues with effective visibilities

petrochenkov opened the pull request:
Continuation of https://github.com/rust-lang/rust/pull/103965.
See individual commits for more detailed description of the changes.

The shortcuts removed in https://github.com/rust-lang/rust/pull/104602/commits/4eb63f618e601efee657d24cd4e8833fb03fac4c and https://github.com/rust-lang/rust/pull/104602/commits/c7c7d1672739e38c8d39ae861b284486aefd5b48 could actually be correct (or correct after some tweaks), but they used global reasoning like "we can skip this update because if the code compiles then some other update should do the same thing eventually".
I have some expertise in this area, but I still have doubt whether such global reasoning was correct or not, especially in presence of all possible exotic cases with imports.
After this PR all table changes should be "locally correct" after every update, even if it may be overcautious.
If similar optimizations are introduced again they will need detailed comments explaining why it's legal to do what they do and providing proofs.

Fixes https://github.com/rust-lang/rust/issues/104249.
Fixes https://github.com/rust-lang/rust/issues/104539.

petrochenkov committed:
effective visibility: Satisfy borrow checker to use resolver lazily from a closure

petrochenkov commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 9d70899ddc863e2b3d3088dcd36a6ec456aea7a5 with merge 2538668a84d818163fa59473d73ecfc72e8dddc2...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9d70899ddc863e2b3d3088dcd36a6ec456aea7a5","merge_sha":"2538668a84d818163fa59473d73ecfc72e8dddc2"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3503010195/jobs/5867764436) [(plain)](https://github.com/rust-lang/rust/commit/9d70899ddc863e2b3d3088dcd36a6ec456aea7a5/checks/9590564966/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling memchr v2.5.0
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling compiler_builtins v0.1.82
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error: internal compiler error: compiler/rustc_middle/src/middle/privacy.rs:203:17: fully private item in the table DefId(0:1512 ~ core[3b09]::intrinsics::{use#11}): Restricted(DefId(0:1495 ~ core[3b09]::intrinsics))
thread 'rustc' panicked at 'Box<dyn Any>', /checkout/compiler/rustc_errors/src/lib.rs:968:33
stack backtrace:
stack backtrace:
   0:     0x7f24486264a2 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h6f63aca0d818ff3d
   1:     0x7f2448694608 - core::fmt::write::h0cd9f5419c66d611
   2:     0x7f24486170c1 - std::io::Write::write_fmt::hdbe174a0534139e1
   3:     0x7f2448626265 - std::sys_common::backtrace::print::h8337b838189f3756
   4:     0x7f2448629577 - std::panicking::default_hook::{{closure}}::h40a27f353e96fa92
   5:     0x7f24486292d5 - std::panicking::default_hook::h7a5f0283402423a0
   6:     0x7f2448fe5b24 - rustc_driver[33805c906714b4ec]::DEFAULT_HOOK::{closure#0}::{closure#0}
   7:     0x7f2448629e83 - std::panicking::rust_panic_with_hook::hec8edde2808e982b
   8:     0x7f244ba3c7c3 - std[f70911762a7dc09b]::panicking::begin_panic::<rustc_errors[f00c99bb6f3a048b]::ExplicitBug>::{closure#0}
   9:     0x7f244ba3c426 - std[f70911762a7dc09b]::sys_common::backtrace::__rust_end_short_backtrace::<std[f70911762a7dc09b]::panicking::begin_panic<rustc_errors[f00c99bb6f3a048b]::ExplicitBug>::{closure#0}, !>
  10:     0x7f2448f88526 - std[f70911762a7dc09b]::panicking::begin_panic::<rustc_errors[f00c99bb6f3a048b]::ExplicitBug>
  11:     0x7f244ba3c416 - std[f70911762a7dc09b]::panic::panic_any::<rustc_errors[f00c99bb6f3a048b]::ExplicitBug>
  12:     0x7f244ba353f1 - <rustc_errors[f00c99bb6f3a048b]::HandlerInner>::span_bug::<rustc_span[8615aa9fcc15d8c5]::span_encoding::Span, &alloc[a6df30274c58c997]::string::String>
  13:     0x7f244ba35207 - <rustc_errors[f00c99bb6f3a048b]::Handler>::span_bug::<rustc_span[8615aa9fcc15d8c5]::span_encoding::Span, &alloc[a6df30274c58c997]::string::String>
  14:     0x7f244bba7a14 - rustc_middle[7be20166cc633508]::ty::context::tls::with_context_opt::<rustc_middle[7be20166cc633508]::ty::context::tls::with_opt<rustc_middle[7be20166cc633508]::util::bug::opt_span_bug_fmt<rustc_span[8615aa9fcc15d8c5]::span_encoding::Span>::{closure#0}, ()>::{closure#0}, ()>
  15:     0x7f244bba8b89 - rustc_middle[7be20166cc633508]::util::bug::opt_span_bug_fmt::<rustc_span[8615aa9fcc15d8c5]::span_encoding::Span>
  16:     0x7f2448f8e1c7 - rustc_middle[7be20166cc633508]::util::bug::span_bug_fmt::<rustc_span[8615aa9fcc15d8c5]::span_encoding::Span>
  17:     0x7f244bbc9154 - <rustc_middle[7be20166cc633508]::middle::privacy::EffectiveVisibilities>::check_invariants
  18:     0x7f24496ec58b - rustc_privacy[6a86c43375f77618]::effective_visibilities
  19:     0x7f244ac99d8a - rustc_query_system[be8adce5b7c83ed2]::query::plumbing::try_execute_query::<rustc_query_impl[d94912f1f7b2d54e]::plumbing::QueryCtxt, rustc_query_system[be8adce5b7c83ed2]::query::caches::DefaultCache<(), &rustc_middle[7be20166cc633508]::middle::privacy::EffectiveVisibilities>>
  20:     0x7f244ad44eb0 - rustc_query_system[be8adce5b7c83ed2]::query::plumbing::get_query::<rustc_query_impl[d94912f1f7b2d54e]::queries::effective_visibilities, rustc_query_impl[d94912f1f7b2d54e]::plumbing::QueryCtxt>
  21:     0x7f244a9256fa - <rustc_query_impl[d94912f1f7b2d54e]::Queries as rustc_middle[7be20166cc633508]::ty::query::QueryEngine>::effective_visibilities
  22:     0x7f2449f697eb - rustc_passes[666026c91860f341]::stability::check_unused_or_stable_features
  23:     0x7f244910a5d0 - <rustc_session[1ff47ef0d4d6d0ce]::session::Session>::time::<(), rustc_interface[e7d8156fe52b0687]::passes::analysis::{closure#0}::{closure#2}::{closure#0}>
  24:     0x7f24491030e5 - std[f70911762a7dc09b]::panic::catch_unwind::<core[867cfca19013d5a]::panic::unwind_safe::AssertUnwindSafe<rustc_interface[e7d8156fe52b0687]::passes::analysis::{closure#0}::{closure#2}>, ()>
  25:     0x7f244910b2c6 - <rustc_session[1ff47ef0d4d6d0ce]::session::Session>::time::<(), rustc_interface[e7d8156fe52b0687]::passes::analysis::{closure#0}>
  26:     0x7f2449132186 - rustc_interface[e7d8156fe52b0687]::passes::analysis
  27:     0x7f244ac91fc0 - rustc_query_system[be8adce5b7c83ed2]::query::plumbing::try_execute_query::<rustc_query_impl[d94912f1f7b2d54e]::plumbing::QueryCtxt, rustc_query_system[be8adce5b7c83ed2]::query::caches::DefaultCache<(), core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>>>
  28:     0x7f244ad6f200 - rustc_query_system[be8adce5b7c83ed2]::query::plumbing::get_query::<rustc_query_impl[d94912f1f7b2d54e]::queries::analysis, rustc_query_impl[d94912f1f7b2d54e]::plumbing::QueryCtxt>
  29:     0x7f244a8f7aca - <rustc_query_impl[d94912f1f7b2d54e]::Queries as rustc_middle[7be20166cc633508]::ty::query::QueryEngine>::analysis
  30:     0x7f244903b788 - <rustc_interface[e7d8156fe52b0687]::passes::QueryContext>::enter::<rustc_driver[33805c906714b4ec]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>>
  31:     0x7f244904f77b - <rustc_interface[e7d8156fe52b0687]::interface::Compiler>::enter::<rustc_driver[33805c906714b4ec]::run_compiler::{closure#1}::{closure#2}, core[867cfca19013d5a]::result::Result<core[867cfca19013d5a]::option::Option<rustc_interface[e7d8156fe52b0687]::queries::Linker>, rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>>
  32:     0x7f2448fe71e2 - rustc_span[8615aa9fcc15d8c5]::with_source_map::<core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>, rustc_interface[e7d8156fe52b0687]::interface::run_compiler<core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>, rustc_driver[33805c906714b4ec]::run_compiler::{closure#1}>::{closure#0}::{closure#1}>
  33:     0x7f2449042ccc - <scoped_tls[b64ea83672690cf8]::ScopedKey<rustc_span[8615aa9fcc15d8c5]::SessionGlobals>>::set::<rustc_interface[e7d8156fe52b0687]::interface::run_compiler<core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>, rustc_driver[33805c906714b4ec]::run_compiler::{closure#1}>::{closure#0}, core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>>
  34:     0x7f244903eaca - std[f70911762a7dc09b]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[e7d8156fe52b0687]::util::run_in_thread_pool_with_globals<rustc_interface[e7d8156fe52b0687]::interface::run_compiler<core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>, rustc_driver[33805c906714b4ec]::run_compiler::{closure#1}>::{closure#0}, core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>>
  35:     0x7f2449048096 - std[f70911762a7dc09b]::panic::catch_unwind::<core[867cfca19013d5a]::panic::unwind_safe::AssertUnwindSafe<<std[f70911762a7dc09b]::thread::Builder>::spawn_unchecked_<rustc_interface[e7d8156fe52b0687]::util::run_in_thread_pool_with_globals<rustc_interface[e7d8156fe52b0687]::interface::run_compiler<core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>, rustc_driver[33805c906714b4ec]::run_compiler::{closure#1}>::{closure#0}, core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>>::{closure#1}::{closure#0}>, core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>>
  36:     0x7f2448ff4339 - <<std[f70911762a7dc09b]::thread::Builder>::spawn_unchecked_<rustc_interface[e7d8156fe52b0687]::util::run_in_thread_pool_with_globals<rustc_interface[e7d8156fe52b0687]::interface::run_compiler<core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>, rustc_driver[33805c906714b4ec]::run_compiler::{closure#1}>::{closure#0}, core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[867cfca19013d5a]::result::Result<(), rustc_errors[f00c99bb6f3a048b]::ErrorGuaranteed>>::{closure#1} as core[867cfca19013d5a]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  37:     0x7f2448636d9e - std::sys::unix::thread::Thread::new::thread_start::h8e0f0b30fe6295ff
  38:     0x7f24483cbb43 - <unknown>
  39:     0x7f244845da00 - <unknown>
  40:                0x0 - <unknown>
note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md


note: rustc 1.67.0-nightly (45f885284 2022-11-19) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 -C debuginfo=0 -C debug-assertions=on -Z unstable-options -C symbol-mangling-version=legacy -Z unstable-options -Z unstable-options -Z macro-backtrace -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -C split-debuginfo=off -C prefer-dynamic -C llvm-args=-import-instr-limit=10 -C embed-bitcode=yes -Z crate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/") -Z binary-dep-depinfo -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
query stack during panic:
#0 [effective_visibilities] checking effective visibilities
#1 [analysis] running analysis passes on this crate
error: could not compile `core`
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:03:49
```

</details>

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3503011160/jobs/5868515360)
Build commit: 2538668a84d818163fa59473d73ecfc72e8dddc2 (`2538668a84d818163fa59473d73ecfc72e8dddc2`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3503011160/jobs/5868515360"},"merge_sha":"2538668a84d818163fa59473d73ecfc72e8dddc2"} -->

rust-timer commented:
Queued 2538668a84d818163fa59473d73ecfc72e8dddc2 with parent 2f8d8040166a730d0da7bba0f2864f0ef7ff6364, future [comparison URL](https://perf.rust-lang.org/compare.html?start=2f8d8040166a730d0da7bba0f2864f0ef7ff6364&end=2538668a84d818163fa59473d73ecfc72e8dddc2).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (2538668a84d818163fa59473d73ecfc72e8dddc2): [comparison URL](https://perf.rust-lang.org/compare.html?start=2f8d8040166a730d0da7bba0f2864f0ef7ff6364&end=2538668a84d818163fa59473d73ecfc72e8dddc2&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=2f8d8040166a730d0da7bba0f2864f0ef7ff6364&end=2538668a84d818163fa59473d73ecfc72e8dddc2&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.2%, 0.3%]   | 10    |
| Regressions ❌ <br /> (secondary)  | 2.0%  | [0.3%, 3.6%]   | 19    |
| Improvements ✅ <br /> (primary)   | -0.9% | [-1.0%, -0.7%] | 9     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.3% | [-1.0%, 0.3%]  | 19    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2f8d8040166a730d0da7bba0f2864f0ef7ff6364&end=2538668a84d818163fa59473d73ecfc72e8dddc2&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.9%  | [2.6%, 3.3%]   | 10    |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.4%, -0.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -3.6% | [-6.1%, -1.1%] | 5     |
| All ❌✅ (primary)                 | -0.4% | [-0.4%, -0.4%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=2f8d8040166a730d0da7bba0f2864f0ef7ff6364&end=2538668a84d818163fa59473d73ecfc72e8dddc2&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 5.2% | [4.4%, 6.7%] | 12    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

petrochenkov committed:
effective visibility: Remove questionable optimizations

First, they require eagerly calculating private visibility (current normal module), which is somewhat expensive.
Private visibilities are also lost once calculated, instead of being cached in the table.

Second, I cannot prove that the optimizations are correct.
Maybe they can be partially reinstated in the future in cases when it's cheap and provably correct to do them.
They will also probably be merged into `fn update` in that case.

Partially fixes https://github.com/rust-lang/rust/issues/104249
Fixes https://github.com/rust-lang/rust/issues/104539

petrochenkov committed:
effective visibility: Fix private visibility calculation for modules

Optimizations removed in the previous commit required this function to behave incorrectly, but now those optimizations are gone so we can fix the bug.

Fixes https://github.com/rust-lang/rust/issues/104249

petrochenkov committed:
effective visibility: Always add table entries for nodes used as parents

Previously if the parent was not in the table, and there was nothing to inherit from, the child's private visibility was used, but that's not correct - the parent may have a larger visibility so we should set it to at least the parent's private visibility.
That parent's private visibility is also inserted into the table for caching, so it's not recalculated later if used again.

petrochenkov commented:
@bors try @rust-timer queue
I'll add https://github.com/rust-lang/rust/pull/103688 on top of this PR if the performance is not good enough.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit c7c7d1672739e38c8d39ae861b284486aefd5b48 with merge 22c1292b3b7f95dc90d94f455d9e747fce27dbe4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c7c7d1672739e38c8d39ae861b284486aefd5b48","merge_sha":"22c1292b3b7f95dc90d94f455d9e747fce27dbe4"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3534640792/jobs/5933810114)
Build commit: 22c1292b3b7f95dc90d94f455d9e747fce27dbe4 (`22c1292b3b7f95dc90d94f455d9e747fce27dbe4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3534640792/jobs/5933810114"},"merge_sha":"22c1292b3b7f95dc90d94f455d9e747fce27dbe4"} -->

rust-timer commented:
Queued 22c1292b3b7f95dc90d94f455d9e747fce27dbe4 with parent 80b3c6dbde3ff89a44f8eaa63e08054398b30ecd, future [comparison URL](https://perf.rust-lang.org/compare.html?start=80b3c6dbde3ff89a44f8eaa63e08054398b30ecd&end=22c1292b3b7f95dc90d94f455d9e747fce27dbe4).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (22c1292b3b7f95dc90d94f455d9e747fce27dbe4): [comparison URL](https://perf.rust-lang.org/compare.html?start=80b3c6dbde3ff89a44f8eaa63e08054398b30ecd&end=22c1292b3b7f95dc90d94f455d9e747fce27dbe4&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=80b3c6dbde3ff89a44f8eaa63e08054398b30ecd&end=22c1292b3b7f95dc90d94f455d9e747fce27dbe4&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.2%, 0.3%]   | 10    |
| Regressions ❌ <br /> (secondary)  | 2.0%  | [0.1%, 3.7%]   | 20    |
| Improvements ✅ <br /> (primary)   | -1.2% | [-1.5%, -0.6%] | 12    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.5% | [-1.5%, 0.3%]  | 22    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=80b3c6dbde3ff89a44f8eaa63e08054398b30ecd&end=22c1292b3b7f95dc90d94f455d9e747fce27dbe4&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.8% | [0.8%, 0.8%] | 1     |
| Regressions ❌ <br /> (secondary)  | 3.6% | [2.4%, 4.4%] | 12    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.8% | [0.8%, 0.8%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=80b3c6dbde3ff89a44f8eaa63e08054398b30ecd&end=22c1292b3b7f95dc90d94f455d9e747fce27dbe4&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 4.7%  | [3.6%, 6.1%]   | 12    |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.5% | [-2.5%, -2.5%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

petrochenkov committed:
effective visibility: Stop recalculating current private visibility

It becomes relatively expensive if done often and shows up during perf profiling.

petrochenkov commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 47cd844468a45ea3dde60b2d19c282639e2e44fc with merge d275439274b5b544b04409b58bdae60e2d5afdbc...
<!-- homu: {"type":"TryBuildStarted","head_sha":"47cd844468a45ea3dde60b2d19c282639e2e44fc","merge_sha":"d275439274b5b544b04409b58bdae60e2d5afdbc"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3536198293/jobs/5936439947)
Build commit: d275439274b5b544b04409b58bdae60e2d5afdbc (`d275439274b5b544b04409b58bdae60e2d5afdbc`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3536198293/jobs/5936439947"},"merge_sha":"d275439274b5b544b04409b58bdae60e2d5afdbc"} -->

rust-timer commented:
Queued d275439274b5b544b04409b58bdae60e2d5afdbc with parent d121aa3b5584eb919a4aaf64dbae0ea1e8e30231, future [comparison URL](https://perf.rust-lang.org/compare.html?start=d121aa3b5584eb919a4aaf64dbae0ea1e8e30231&end=d275439274b5b544b04409b58bdae60e2d5afdbc).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (d275439274b5b544b04409b58bdae60e2d5afdbc): [comparison URL](https://perf.rust-lang.org/compare.html?start=d121aa3b5584eb919a4aaf64dbae0ea1e8e30231&end=d275439274b5b544b04409b58bdae60e2d5afdbc&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=d121aa3b5584eb919a4aaf64dbae0ea1e8e30231&end=d275439274b5b544b04409b58bdae60e2d5afdbc&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.2%, 0.3%]   | 6     |
| Regressions ❌ <br /> (secondary)  | 1.8%  | [0.2%, 3.5%]   | 20    |
| Improvements ✅ <br /> (primary)   | -1.1% | [-1.4%, -0.5%] | 13    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.7% | [-1.4%, 0.3%]  | 19    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d121aa3b5584eb919a4aaf64dbae0ea1e8e30231&end=d275439274b5b544b04409b58bdae60e2d5afdbc&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.4%  | [2.4%, 2.4%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 3.2%  | [2.6%, 4.0%]   | 12    |
| Improvements ✅ <br /> (primary)   | -1.3% | [-1.3%, -1.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.6% | [-2.6%, -1.0%] | 3     |
| All ❌✅ (primary)                 | 0.5%  | [-1.3%, 2.4%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d121aa3b5584eb919a4aaf64dbae0ea1e8e30231&end=d275439274b5b544b04409b58bdae60e2d5afdbc&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 4.3% | [3.0%, 5.9%] | 11    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

petrochenkov commented:
I'm not a fan of mixing correctness and performance changes, but I've added one optimization commit on top of this PR, and it didn't change much. I can remove it again if requested.
[unused-warnings](https://github.com/rust-lang/rustc-perf/tree/master/collector/compile-benchmarks/unused-warnings) is basically a stress test for this pass - it contains tens of thousands of import chains.

r? @oli-obk @Bryanskiy @cjgillot 
@rustbot ready

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 47cd844468a45ea3dde60b2d19c282639e2e44fc has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 47cd844468a45ea3dde60b2d19c282639e2e44fc -->
<!-- homu: {"type":"Approved","sha":"47cd844468a45ea3dde60b2d19c282639e2e44fc","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 47cd844468a45ea3dde60b2d19c282639e2e44fc with merge 41e0363055ade59584cff667c79f64937e6ef3f9...
<!-- homu: {"type":"BuildStarted","head_sha":"47cd844468a45ea3dde60b2d19c282639e2e44fc","merge_sha":"41e0363055ade59584cff667c79f64937e6ef3f9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3545768513/jobs/5956083873)
Approved by: oli-obk
Pushing 41e0363055ade59584cff667c79f64937e6ef3f9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3545768513/jobs/5956083873"},"merge_sha":"41e0363055ade59584cff667c79f64937e6ef3f9"} -->

rust-timer commented:
Finished benchmarking commit (41e0363055ade59584cff667c79f64937e6ef3f9): [comparison URL](https://perf.rust-lang.org/compare.html?start=af63e3b39f3990418ad8e0a1b1fa8a722a7c50b0&end=41e0363055ade59584cff667c79f64937e6ef3f9&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=af63e3b39f3990418ad8e0a1b1fa8a722a7c50b0&end=41e0363055ade59584cff667c79f64937e6ef3f9&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.2%, 0.3%]   | 8     |
| Regressions ❌ <br /> (secondary)  | 2.2%  | [0.3%, 3.6%]   | 17    |
| Improvements ✅ <br /> (primary)   | -1.0% | [-1.4%, -0.5%] | 13    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.5% | [-1.4%, 0.3%]  | 21    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=af63e3b39f3990418ad8e0a1b1fa8a722a7c50b0&end=41e0363055ade59584cff667c79f64937e6ef3f9&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.0%  | [2.2%, 4.1%]   | 10    |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.5% | [-3.5%, -3.5%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=af63e3b39f3990418ad8e0a1b1fa8a722a7c50b0&end=41e0363055ade59584cff667c79f64937e6ef3f9&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 4.0%  | [3.2%, 5.0%]   | 12    |
| Improvements ✅ <br /> (primary)   | -1.2% | [-1.4%, -1.0%] | 3     |
| Improvements ✅ <br /> (secondary) | -2.6% | [-2.6%, -2.6%] | 2     |
| All ❌✅ (primary)                 | -1.2% | [-1.4%, -1.0%] | 3     |
</details>

nnethercote commented:
This fixes correctness issues, and the regressions are mostly in `unused-warnings` which is a stress test where 3% doesn't really matter. Plus `libc` got a bit faster. So overall, not a performance concern.

@rustbot label: +perf-regression-triaged
